src/vcf/fixer.cpp

```diff
@@ -14,385 +14,582 @@
+                format_fields[format_keys[i]].push_back(i);
+                if (format_fields[format_keys[i]].size() == 1) {
+                    ordered_fields.push_back(format_keys[i]);
+                }
             }
 
-            // write the FORMAT column
-            output << *first;
+            std::set<std::string> fields_to_remove;
+
+            for (auto & format_field : format_fields) {
+                if (format_field.second.size() > 1) {
+                    for (auto it = first + 1; it != last; it++) {
+                        std::vector<std::string> sample_fields;
+                        util::string_split(*it, ":", sample_fields);
+                        for (auto & format_field_index : format_field.second) {
+                            if (sample_fields.size() > format_field_index && sample_fields[format_field_index] != sample_fields[format_field.second[0]]) {
+                                fields_to_remove.insert(format_field.first);
+                                it = last - 1;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
 
-            // `cardinality` should be -1 if the cardinality is unknown, and any positive number otherwise
-            // so, if unknown, put 1, so that a single "." is written
-            size_t repeat = error.field_cardinality <= -1 ? 1 : static_cast<size_t>(error.field_cardinality);
+            std::string fixed_format;
+            std::vector<std::string> fixed_samples(last - (first + 1));
+
+            for (auto & ordered_field : ordered_fields) {
+                if (fields_to_remove.find(ordered_field) == fields_to_remove.end()) {
+                    // keep first occurrence, discard the rest if present
+                    fixed_format += ordered_field + ":";
+                    for (auto it = first + 1; it != last; it++) {
+                        std::vector<std::string> sample_fields;
+                        util::string_split(*it, ":", sample_fields);
+                        if (sample_fields.size() > format_fields[ordered_field][0]) {
+                            fixed_samples[it - (first + 1)] += sample_fields[format_fields[ordered_field][0]] + ":";
+                        }
+                    }
+                } else {
+                    num_removed_duplicates++;
+                }
+            }
 
-            // now `it` will point to each SAMPLE column
-            for (auto it = ++first; it != last; ++it) {
-                output << "\t";
-                fix_column(gt_column_index, *it, field_separator, [&](std::string wrong_subfield) {
-                    util::print_container(output, std::string(repeat, empty_subfield), "", subfield_separator, "");
-                });
+            if (fixed_format.empty()) {
+                fixed_format = MISSING_VALUE;         // all format fields were removed, so use missing value
```

Comment:
I guess FORMAT does not permit a single dot `.`, we also throw an error in the validator. So in this case, where the Format column must be empty, should we throw an exception or keep it as MISSING_VALUE?